### PR TITLE
Making package solver agnostic

### PR DESCRIPTION
**Pull request by michel2323**
A first step in removing solvers.

### CODE REVIEW

**Comment by michel2323** [comment]
@AnirudhSubramanyam Am I setting the options correctly in `runtests.jl`. The tests are passed.

File: src/ProxAL.jl [hunk 1]
```diff
@@ -18,10 +18,10 @@ include("../src/proxALMutil.jl")
 export RawData, ModelParams, AlgParams
 export opf_loaddata, solve_fullmodel, run_proxALM, set_rho!
 
-function run_proxALM(opfdata::OPFData, rawdata::RawData;
+function run_proxALM(opfdata::OPFData, rawdata::RawData, optimizer;
```
Comment (AnirudhSubramanyam) [inline src/ProxAL.jl:21]: Can we include the `optimizer` as part of AlgParams? That would probably be the right place to put that.

File: src/opfblocks.jl [hunk 2]
```diff
@@ -5,12 +5,13 @@ mutable struct OPFBlockData
     blkCount::Int64
     blkIndex::CartesianIndices
     blkModel::Vector{JuMP.Model}
+    blkOptimizer
```
Comment (AnirudhSubramanyam) [inline src/opfblocks.jl:8]: we can probably remove this `blkOptimizer` and use the `algparams.optimizer`

File: src/ProxAL.jl [hunk 3]
```diff
@@ -18,10 +18,10 @@ include("../src/proxALMutil.jl")
...
                      modelinfo::ModelParams = ModelParams(),
                      algparams::AlgParams = AlgParams())
-    runinfo = ProxALMData(opfdata, rawdata; modelinfo = modelinfo, algparams = algparams, fullmodel = true)
+    runinfo = ProxALMData(opfdata, rawdata, optimizer; modelinfo = modelinfo, algparams = algparams, fullmodel = true)
```
Comment (AnirudhSubramanyam) [inline src/ProxAL.jl:24]: Can we include the `optimizer` as part of AlgParams? That would probably be the right place to put that.

File: src/ProxAL.jl (hunk 1 above)
Comment (AnirudhSubramanyam) [inline src/ProxAL.jl:21]: Can we include the `optimizer` as part of AlgParams? That would probably be the right place to put that.

File: src/opfblocks.jl [hunk 4]
```diff
@@ -5,12 +5,13 @@ mutable struct OPFBlockData
...
     blkOpfdt::Vector{OPFData}
     blkMInfo::Vector{ModelParams}
     colCount::Int64
     colValue::Array{Float64,2}
 
-    function OPFBlockData(opfdata::OPFData, rawdata::RawData;
+    function OPFBlockData(opfdata::OPFData, rawdata::RawData, optimizer;
```
Comment (AnirudhSubramanyam) [inline src/opfblocks.jl:14]: Can we include the `optimizer` as part of AlgParams? That would probably be the right place to put that.

File: src/opfblocks.jl [hunk 5]
```diff
@@ -58,17 +59,15 @@ mutable struct OPFBlockData
...
 end
 
 function opf_block_model_initialize(blk::Int, opfblocks::OPFBlockData, rawdata::RawData;
                                     algparams::AlgParams)
     @assert blk >= 1 && blk <= opfblocks.blkCount
 
-    opfmodel = JuMP.Model(optimizer_with_attributes(Ipopt.Optimizer,
-                                                    "print_level" => Int64(algparams.verbose > 0)*2,
-                                                    "max_iter" => algparams.nlpiterlim))
+    opfmodel = JuMP.Model(opfblocks.blkOptimizer)
```
Comment (AnirudhSubramanyam) [inline src/opfblocks.jl:70]: This becomes `algparams.optimizer`

**Review by AnirudhSubramanyam — CHANGES_REQUESTED** [review]
Can we include `optimizer` as part of AlgParams, instead of passing it as an additional argument? That would probably be the right place to put that. Also, we can remove the `OPFBlockData.blkOptimizer` object and use the `AlgParams.optimizer` instead inside the NLP blocks.

**Comment by michel2323** [comment]
> Can we include `optimizer` as part of AlgParams, instead of passing it as an additional argument? That would probably be the right place to put that. Also, we can remove the `OPFBlockData.blkOptimizer` object and use the `AlgParams.optimizer` instead inside the NLP blocks.

That sounds good. Will get into it.

**Comment by michel2323** [comment]
@AnirudhSubramanyam  

* I put two solver fields into AlgParams.
* There are a lot of optional keyword arguments that shouldn't be optional as they can be uninitialized. I fixed most of them, but I am not sure whether this maps to your algorithm. Aka, when to switch algorithms and set their parameters.
* IMO we should merge this and continue to work both on this. I plan to work on the solver interfaces. MadNLP is currently broken due to some ambiguity in a `copy!` call, but Sungho is on it. HiOp has a whole lot of issues :-) .

File: src/opfblocks.jl (hunk 5 above)
Comment (michel2323) [reply src/opfblocks.jl:70]: This is addressed by the latest commit.

**Review by AnirudhSubramanyam — APPROVED** [review]

**Pull request merged**